graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/caching/CacheControlInstrumentation.java

```diff
@@ -0,0 +1,344 @@
+  private final int defaultMaxAge;
+
+  private static final Object CONTEXT_KEY = new Object();
+  private static final String DIRECTIVE_NAME = "cacheControl";
+  private static final String MAX_AGE = "maxAge";
+  private static final String SCOPE = "scope";
+  private static final String INHERIT_MAX_AGE = "inheritMaxAge";
+
+  public CacheControlInstrumentation() {
+    this(0);
+  }
+
+  public CacheControlInstrumentation(int defaultMaxAge) {
+    this.defaultMaxAge = defaultMaxAge;
+  }
+
+  @Nullable
+  public static String cacheControlContext(GraphQLContext context) {
+    return context.get(CONTEXT_KEY);
+  }
+
+  @Override
+  public InstrumentationState createState() {
+    return new CacheControlState();
+  }
+
+  @Override
+  public InstrumentationContext<ExecutionResult> beginExecution(
+      InstrumentationExecutionParameters parameters) {
+    return new InstrumentationContext<ExecutionResult>() {
+      @Override
+      public void onDispatched(CompletableFuture<ExecutionResult> completableFuture) {}
+
+      @Override
+      public void onCompleted(ExecutionResult executionResult, Throwable throwable) {
+        CacheControlState state = parameters.getInstrumentationState();
+
+        // Attach the policy to the context object
+        state
+            .overallPolicy
+            .maybeAsString()
+            .ifPresent(s -> parameters.getGraphQLContext().put(CONTEXT_KEY, s));
+      }
+    };
+  }
+
+  @Override
+  public InstrumentationContext<ExecutionResult> beginField(
+      InstrumentationFieldParameters parameters) {
+    CacheControlState state = parameters.getInstrumentationState();
+    CacheControlPolicy fieldPolicy = new CacheControlPolicy();
+    boolean inheritMaxAge = false;
+
+    GraphQLUnmodifiedType unwrappedFieldType =
+        GraphQLTypeUtil.unwrapAll(parameters.getExecutionStepInfo().getType());
+
+    // There's no way to set a cacheControl directive on the _entities field or
+    // the _Entity union in SDL. Instead, we can determine the possible concrete
+    // types from the representations arguments and select the most restrictive
+    // cache policy from those types.
+
+    if (unwrappedFieldType.getName().equals(_Entity.typeName)) {
+      Object representations = parameters.getExecutionStepInfo().getArgument(_Entity.argumentName);
+
+      if (representations instanceof List) {
+        typesFromEntitiesArgument(
+                representations, parameters.getExecutionContext().getGraphQLSchema())
+            .stream()
+            .map(
+                type ->
+                    CacheControlDirective.fromDirectiveContainer((GraphQLDirectiveContainer) type))
+            .filter(Optional::isPresent)
+            .forEach(directive -> fieldPolicy.restrict(directive.get()));
+      }
+    } else if (unwrappedFieldType instanceof GraphQLCompositeType) {
+
+      // Cache directive on the return type of this field if it's a composite type
+
+      Optional<CacheControlDirective> directive =
```

Comment:
I'm not sure I understand the logic of what you're still using Optional for vs not (I see it's no longer in fields or params which makes sense), but if @martinbonnin is happy with this level of use then I won't complain

Reply:
I'd personally ditch `Optional` altogether but don't mind it too much and as long as it's not public API, I think it's fine to ship like this and we can always refine later if/when we want to.